### DISCUSSION TITLE
Intersection of multiple geometries

### DISCUSSION THREAD

**Discussion by MaroonMike1** [NetTopologySuite/NetTopologySuite#544] [Q&A] [answered]
Newbie question, but I am looking to get the distinct intersection of 3 (or more) polygons that overlap.  I see that I can use Feature.Geometry.Intersection() to compare two geometries, but is there a method to return the unique intersection (a geometry) of multiple geometries as input?

**Comment by FObermaier** [comment, marked as answer]
No, there is no such -possibly optimized- functionality. You have to do it sequentially, sth. along these lines:
```C#
// Warning: Has not seen a compiler
public static void Geometry Intersection(IList<Geometry> geoms) {
    var res = geoms[0];
    for (int i = 1; i < geoms.Count; i++) {
        res = res.Intersection(geoms[i]); 
        if (res.IsEmpty) break;
    }
    return res;
}
```

**Reply by airbreather** [reply]
Because Intersection is commutative and associative like Union is, you could also do this by building a tree, useful if there are **many**.  Somewhat naïve approach:
```csharp
// Warning: also has not seen a compiler
// returns null if the list is empty OR if the intersection is empty.
// can be modified to be guaranteed to return an empty geometry if the list is non-empty.
public static Geometry Intersection(IReadOnlyList<Geometry> geoms)
{
    var curr = new List<Geometry>(geoms);
    var next = new List<Geometry>();
    while (curr.Count > 1)
    {
        for (int i = 0; i < curr.Count; i += 2)
        {
            var geom = curr[i];
            if (i + 1 < curr.Count)
            {
                geom = geom.Intersection(curr[i + 1]);
            }

            if (!geom.IsEmpty)
            {
                next.Add(geom);
            }
        }

        (curr, next) = (next, curr);
        next.Clear();
    }

    return next.SingleOrDefault();
}
```
**PERHAPS** better would be to follow how `UnaryUnionOp` / `CascadedPolygonUnion` works by using a `STRtree<Geometry>` to group nearby inputs together, but that's going to be a lot more code than I can put in a quick sample like this :wink:

**Reply by MaroonMike1** [reply]
This was very helpful and resolved my issue.  Thank you.
